clash-cores/src/Clash/Cores/Xilinx/BlockRam/Internal.hs

```diff
@@ -0,0 +1,164 @@
+{-|
+  Copyright   :  (C) 2023 QBayLogic B.V.
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+  Model implementation for primitives in "Clash.Cores.Xilinx.BlockRam".
+-}
+
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Clash.Cores.Xilinx.BlockRam.Internal where
+
+import Clash.Explicit.Prelude hiding (enable)
+import Clash.Explicit.BlockRam.Model (tdpbramModel)
+import Clash.XException.MaybeX (MaybeX(..))
+
+import Clash.Annotations.Primitive (Primitive(InlineYamlPrimitive), hasBlackBox)
+import Clash.Cores.Xilinx.BlockRam.BlackBoxes (tdpbramBBF)
+import Clash.Sized.Internal.BitVector (BitVector(BV), undefined#)
+
+import Data.String.Interpolate (__i)
+import GHC.Stack (HasCallStack)
+
+import qualified Data.Sequence as Seq
+
+-- $setup
+-- >>> import Clash.Explicit.Prelude
+
+
+-- | Returns a byte enable such that it agrees with the global enable of a
+-- block RAM. I.e., a deasserted global enable will return a deasserted byte
+-- enable.
+--
+-- >>> mergeEnableByteEnable @3 (pure undefined) (pure $(bLit "01."))
+-- IsDefined_ 0b0..
+-- >>> mergeEnableByteEnable @3 (pure False) (pure $(bLit "01."))
+-- IsDefined_ 0b000
+-- >>> mergeEnableByteEnable @3 (pure True) (pure $(bLit "01."))
+-- IsDefined_ 0b01.
+-- >>> mergeEnableByteEnable @3 (pure True) (pure undefined)
+-- IsX_ ...
+--
+mergeEnableByteEnable ::
+  KnownNat n =>
+  -- | Global enable
+  MaybeX Bool ->
+  -- | Byte enable
+  MaybeX (BitVector n) ->
+  -- | Byte enable combined with global enable
+  MaybeX (BitVector n)
+mergeEnableByteEnable enaX byteEnaX =
+  case enaX of
+    IsX _           -> liftA2 (.&.) (pure undefined#) byteEnaX
+    IsDefined True  -> byteEnaX
+    IsDefined False -> pure 0
+
+-- | Clone each bit in a mask 8 times.
```

Comment:
I don't know what the deal is with the training dot. Do we care? And if so, we should just enforce it somehow. These comments make reviews a real chore if I'm being honest.

Reply:
You can't enforce it, as far as I'm aware. We end sentence fragments without a period and full sentences in a period. Enforcing it sounds like you'd need a computer program to tell whether something is a sentence fragment or a full sentence. It's just something we agreed to do. We can discuss whether we want it, but I'm inclined to say yes.

Would you prefer if in the future I comment on it once and say "and please take care of all other sentence fragments as well"? I thought it would be useful to know which places to change.

Reply:
You caught me in a weak moment; I do appreciate the scrutiny sorry. On this particular issue I feel very much _meh_, though. There are a couple of reasons:

1. I don't think we've written it down anywhere that we "end sentence fragments without a period and full sentences in a period", nor have a really heard of it before. I also can't find anything on the internet saying that this is a norm. (If anything, TheInternet suggest to entirely avoid sentence fragments in written text. So does our style guide!) This ticked me off more than it should have.
 
2. I don't feel it's particularly important.

3. It feels entirely arbitrary. Couldn't we just decide to end everything with a `.` instead? This side-steps the weird case where you _typically_ not end with a `.`, but you sometimes do because you need two (partial) sentences. This happens often in changelog entries. Plus, it spares brain cycli figuring out whether something is a full sentence or a fragment.

Reply:
The Google style guides do contain some guidance on this, but interestingly, the different languages don't necessarily agree, and none of them clearly support my stance. Some examples:

[Go Style Guide](https://google.github.io/styleguide/go/decisions.html#comment-sentences)
> [Documentation comments](https://google.github.io/styleguide/go/decisions.html#doc-comments) should always be complete sentences, and as such should always be capitalized and punctuated. Simple end-of-line comments (especially for struct fields) can be simple phrases that assume the field name is the subject.

They write their doc as complete sentences. I'm not in favour of this. Note the slight suggestion that sentences that are not complete might not need punctuation (_as such, punctuated_). Could be me reading too much into it.

[Python Style Guide](https://google.github.io/styleguide/pyguide.html#386-punctuation-spelling-and-grammar)
>Pay attention to punctuation, spelling, and grammar; it is easier to read well-written comments than badly written ones.
>
>Comments should be as readable as narrative text, with proper capitalization and punctuation. In many cases, complete sentences are more readable than sentence fragments. Shorter comments, such as comments at the end of a line of code, can sometimes be less formal, but you should be consistent with your style.
>
>Although it can be frustrating to have a code reviewer point out that you are using a comma when you should be using a semicolon, it is very important that source code maintain a high level of clarity and readability. Proper punctuation, spelling, and grammar help with that goal.

Note that they also consider inline comments there, not just documentation...

[Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml?showone=Grammar_and_punctuation#Grammar_and_punctuation)
> You should punctuate documentation correctly.
> When a comment is a full sentence, you should capitalize the initial letter of the first word and end the comment with a period. In general, you should use correct punctuation.

Again leaves the door open for sentence fragments to not end in a period.

I think it looks weird if sentence fragments that are their own paragraph end in a period.

The only advice I see in our own style guide is
> Write proper sentences; start with a capital letter and use proper punctuation.

I'm arguing precisely that a sentence fragment is often proper and that the proper punctuation for that can be that it should not end in a period. I'd like to note that I think, nay, hope even, that the majority of our current `clash-prelude` Haddock does not have a period at the end of a lone sentence fragment. If our stye guide said
> Write full sentences; start with a capital letter and end in a period.

I wouldn't have a leg to stand on, although I'd start rallying for changing that guideline because using full sentences everywhere is so much unnecessary fluff.

It doesn't feel arbitrary to me. I believe we held this discussion when Alex still worked for QBayLogic. It's very well possible you weren't there; I have no idea which group of people it was we discussed it with. If you want to end all our sentences, fragment or not, in a period, I'd like to invite Alex to weigh in, given it's his mother's tongue (well, not completely, since we decided to use American English for our doc).

You don't feel it's important; but it is somewhat important to me. Consistency is really important to me. But I'm also put off by a period at the end of a sentence fragment. So I'd still like a discussion. I can let go of the requirement to omit the period at the end of a sentence fragment; I'll just have to suck it up. But I don't want to let go of consistency.

Reply:
I hear what you're saying.